### ISSUE TITLE
BreakMeshByBlock robustness

### ISSUE THREAD

**Issue by jiangwen84** [idaholab/moose#12033]
## Rationale

There are two main issues with existing BreakMeshByBlock
1. the interface sideset is not assigned correctly.
2. the connectivity of the elements with news nodes are not correct.

## Description

resolve those issues in BreakMeshByBlock.

## Impact

Improvement and bug fixes.

**Comment by lindsayad** [comment]
It seems that `BreakMeshByBlock(Generator)` suffers from a fair number of robustness issues. Based on @arovinelli's comment [here](https://github.com/idaholab/moose/pull/14124#issuecomment-544210278) it does not work with recover (at least when using `InterfaceKernels`?) nor with `DistributedMesh`. Additionally, according to the test suite [it doesn't work in parallel in debug mode either](https://civet.inl.gov/job/414956/). It would be nice if we could make this object a bit more robust such that it can work in all the traditional MOOSE scenarios.

**Comment by arovinelli** [comment]
@lindsayad making it work with distributed mesh should be double even if a bit cumbersome. For the restart I believe the major problem is that the mesh connectivity is not saved there ant it might be difficult to reconstruct it when the mesh is already split (if we have large dispalcements between the two blocks is going to be a impossible). Any idea on how to solve this?

> It seems that `BreakMeshByBlock(Generator)` suffers from a fair number of robustness issues. Based on @arovinelli's comment [here](https://github.com/idaholab/moose/pull/14124#issuecomment-544210278) it does not work with recover (at least when using `InterfaceKernels`?) nor with `DistributedMesh`. Additionally, according to the test suite [it doesn't work in parallel in debug mode either](https://civet.inl.gov/job/414956/). It would be nice if we could make this object a bit more robust such that it can work in all the traditional MOOSE scenarios.

**Comment by permcody** [comment]
Closing this specific issue based on the merged PRs listed.

**Comment by lindsayad** [comment]
Err I don't think this should be closed, #14216 specifically references this issue **because** it is still an issue

**Comment by arovinelli** [comment]
@lindsayad and @permcody  I need to run some big jobs on a cluster and therefore I need to fix at the least to enable the distributed mesh option. I guess that the problem here is to sync different MPI process to add node and update element simultaneously. Can you point me to any example doing something similar to give me an example to follow and simplify my work? Of course if Ica n fix some issue this will be merged.
Thanks again

**Comment by permcody** [comment]
@arovinelli - You can jump around other generators and see if others have been updated. It can be pretty complex.

Here's a better idea. Instead of trying to get it working with DistributedMesh, I recommend that you just use the "split mesh" capability in MOOSE. The short explanation is that you run a job with a few extra CLI options that will just run the mesh steps (reading it in, generating it, applying transformations, etc) and will write it back out as separate files ready to run for a larger distributed run. The nice part is when you do this, the memory used by your larger run and the startup time will be drastically reduced. Take a look here:

https://www.mooseframework.org/syntax/Mesh/splitting.html

**Comment by arovinelli** [comment]
@permcody thanks for the reply this is good to know.
One question would the file written from the mesh split contain the face-face neighbor information, or it will be reconstructed later? If it is reconstructed later than I need a way to reassign the correct neighbors

**Comment by permcody** [comment]
face-face neighbor information? I assume you mean the new internal boundaries you are trying to add?

The splitter system is designed to run all mesh setup tasks including running RelationshipManagers. So if you have Generators that add boundaries that information will be saved. If you have objects that require a "wider stencil", those ghost elements should be written. Ideally everything you need should be preserved.

**Comment by arovinelli** [comment]
@permcody face-face neighbor information I mean elem->face->neighbor_elem information
The BreackMeshByBlockGenerator add nodes such that neighboring element on two different block do not share the same nodes anymore. So the mesh is in fact broken, but the elem->face->neighbor_elem information are  still there. If the elem->face->neighbor_elem info are not saved by the mesh split, I need to reconstruct it be cause otherwise the interface will not work anymore

**Comment by arovinelli** [comment]
@permcody and @lindsayad 
so I tried to use the split mesh tool, howeve when treying use the split mesh things brake
this is the output

``` 
*** Error in `../../../../tensor_mechanics-opt': malloc(): memory corruption (fast): 0x0000000000f35f10 ***
*** Error in `../../../../tensor_mechanics-opt': malloc(): memory corruption (fast): 0x000000000228aff0 ***
*** Error in `../../../../tensor_mechanics-opt': malloc(): memory corruption (fast): 0x00000000024d4270 ***
```

**Comment by permcody** [comment]
This might mean that you don't have enough ghosting saved in the split mesh
file. It's important that each object in the simulation advertises its
individual needs as far as accessing nodes, elements, doc indices etc
outside of its partition. Does this sim run with Distributed Mesh? That's
worth checking before continuing down the Split Mesh path.

**Comment by arovinelli** [comment]
@permcody it will not as 

https://github.com/idaholab/moose/blob/5cb1f32e014b750c4815b85a41310b3d80852929/framework/src/meshgenerators/BreakMeshByBlockGenerator.C#L34-L39

**Comment by permcody** [comment]
OK to be clear, Split Mesh is always for Distributed Mesh runs. We could
probably improve error messaging along those lines. This Generator will
need to be updated to work for DistributedMesh AND include appropriate
RelationshipManager additions before you can expect it to work with split
mesh.

**Comment by arovinelli** [comment]
@permcody thanks for the reply
I believe it would be much easier going the other way around start from a disconnected mesh and join coincident faces (e.g. the one having nodes sharing the same locations). I'll work on that

**Comment by permcody** [comment]
> I believe it would be much easier going the other way around start from a disconnected mesh and join coincident faces (e.g. the one having nodes sharing the same locations).

It shouldn't be more difficult one way or the other. When working with a DistributedMesh in the MeshGenerator system, you'll have access to the full mesh on each processor (well in almost all cases). The way you deal with DistributedMesh is that you inspect the "processor_id" on each element to know whether the current processor owns it or not.

**Comment by arovinelli** [comment]
Ok, I didn't know that. I Have a couple of questions?

1.  when adding a new node, which processor should be in charge of adding the new node? the processor owning the element to which the node will belong? 
2.  Also once a node has been added, how will I broadcast this information to all the processors?

> It shouldn't be more difficult one way or the other. When working with a DistributedMesh in the MeshGenerator system, you'll have access to the full mesh on each processor (well in almost all cases). The way you deal with DistributedMesh is that you inspect the "processor_id" on each element to know whether the current processor owns it or not.

**Comment by permcody** [comment]
During the generation phase you want the mesh to be consistent on all processors. All processors will have to add the same Node with the same IDs. If anything gets out of sync, you'll run into problems. I should say that it is possible to generate a DistributedMesh where each processor truly works with only their portion of the mesh, but that's very advanced and we aren't doing that in more than maybe one place in MOOSE.

If you need to perform parallel communication you may do so. However, hopefully since you are doing the same work on all processors you won't need to worry about this step.

**Comment by arovinelli** [comment]
If all processors have access to all the mesh and all processor must do the same thing, why should I be worried if a node and or element is remote or local? Are there any caveats for remote node and elements?

> During the generation phase you want the mesh to be consistent on all processors. All processors will have to add the same Node with the same IDs. If anything gets out of sync, you'll run into problems. I should say that it is possible to generate a DistributedMesh where each processor truly works with only their portion of the mesh, but that's very advanced and we aren't doing that in more than maybe one place in MOOSE.
> 
> If you need to perform parallel communication you may do so. However, hopefully since you are doing the same work on all processors you won't need to worry about this step.

**Comment by permcody** [comment]
If all processors have access to all the mesh and all processor must do the same thing, why should I be worried if a node and or element is remote or local? Are there any caveats for remote node and elements?

I'll admit that I'm not sure we've fleshed out every single possible case when creating a tree of generators. We are still polishing up this system and trying to create documentation and guidance for everyone to use. My concern is that you might generate a sequence of generators A -> B -> C. It might turn out that you need to call the "prepare_for_use()" method on the Mesh between the stages of the generator because you need to call certain "exploration methods" like retrieving neighboring elements of elements, etc. I'm not an expert at what you can and can't do but many methods aren't available for use until you've done an intermediate prepare. However, with DistributedMesh once you do that, I believe (again need to test and verify) that libMesh will go ahead and throw away remote elements. 

Now you get into one of the later stages and you no longer have all of the elements available on all processors. Now if you plan to add new elements/nodes everything gets significantly more complicated. You no longer need to do the same work on all ranks, but you likely can't just do the work on one rank either. An example would be where you were going to add new elements on a free surface right on a processor boundary. Multiple processors would need to work together to add a new element due to ghosting. 

For now, we've kind of glossed over some of these really nasty edge cases (literal) and just considered the more normal case of having the full mesh available on all ranks during the generation phase. That'll satisfy nearly every case and if you are willing to work with Split Mesh, perhaps 100% of cases. Yeah we are figuring this all out as we go to. It's all pretty new and you are once again working on the bleeding edge.

For now, I would recommend that you assume you have all information on all ranks and just do the same work everywhere. In the future we will continue to think about how to make this easier for our developers and give them the tools they need to deal with truly Distributed cases.

**Comment by lindsayad** [comment]
> However, with DistributedMesh once you do that, I believe (again need to test and verify) that libMesh will go ahead and throw away remote elements.

By default it will, but there is a method to stop this from happening: [`MeshBase::allow_remote_element_removal`](https://libmesh.github.io/doxygen/classlibMesh_1_1MeshBase.html#a005d60012e980f1e3c983d13fcc288bd)

**Comment by permcody** [comment]
> By default it will, but there is a method to stop this from happening: MeshBase::allow_remote_element_removal

True, but is this something we want to always turn on during MeshGeneration? Perhaps, but maybe not. Design designs that just haven't been made.

**Comment by arovinelli** [comment]
@permcody @lindsayad 
so I'm trying to figure out what's going on in this generator with distributed mesh.

I compiled in debug and used valgrind and run with distributed mesh (after commenting the top moose error) and the following line is cause of a memory leak

https://github.com/idaholab/moose/blob/5cb1f32e014b750c4815b85a41310b3d80852929/framework/src/meshgenerators/BreakMeshByBlockGenerator.C#L100

any idea how to fix this?

**Comment by lindsayad** [comment]
> True, but is this something we want to always turn on during MeshGeneration? Perhaps, but maybe not. Design designs that just haven't been made.

It seems like you should only delete remote elements once the mesh has been fully generated. That's certainly the safest. If I'm doing contact for example, with the current design there is no way for me to properly generate a mortar interface (with the proper ghosting) unless I've pre-built the lower dimensional bocks in my file mesh. This is because if I try to add lower dimensional blocks via LowerDBlockFromSidesetGenerator, I'll be unable to generate the proper ghosting between slave and master lower dimensional blocks because more than likely the parents were deleted during `mesh->read` within the `FileMeshGenerator`. (Yes `UnstructuredMesh::read` calls `prepare_for_use`)

**Comment by lindsayad** [comment]
> any idea how to fix this?

Did you add this new node to the mesh? You've performed a dynamic memory allocation here. Somebody has to manage this new resource and make sure the dynamic memory that `new_node` points to gets deleted.

**Comment by arovinelli** [comment]
Yes it is added to the mesh

https://github.com/idaholab/moose/blob/5cb1f32e014b750c4815b85a41310b3d80852929/framework/src/meshgenerators/BreakMeshByBlockGenerator.C#L100-L102

actual even line 102 generate a leak most likely because of line 100

> Did you add this new node to the mesh? You've performed a dynamic memory allocation here. Somebody has to manage this new resource and make sure the dynamic memory that `new_node` points to gets deleted.

**Comment by permcody** [comment]
libMesh is still working on updating its interfaces to C++11 standards. Allocating a new node and adding it to the mesh is the way it's been done since libMesh was created. Now in this intermediate stage, it looks very weird since there's a mix of smart and non-smart pointer interfaces. The creation of a new Node through a smart pointer requires that we release the manager before adding it to the old (manually managed) interface. As far as I can tell, this is all correct. libMesh expects you to pass it a Node (or Element) through a pointer that it will then manage.

**Comment by lindsayad** [comment]
Yea, I'm gonna be honest: I don't see anything that you're doing wrong here. Even if the new node's proc_id doesn't match the current rank, `DistributedMesh` should be properly deleting it during `delete_remote_elements`. Maybe @roystgnr has some ideas

**Comment by YaqiWang** [comment]
I actually do not quite like the fact of libMesh having `DistributedMesh` and `ReplicatedMesh`. A simple flag indicating if the mesh is distributed should work imo. But this is a design choice made long ago, I guess we have no way to change it. We can also separate the remote element removal out from `prepare_for_use` and call it `distribute`. That might be helpful. BTW, Rattlesnake has one similar modifier for breaking meshes. I am wondering if that one has the similar issue or not. We used that one for our multi-scheme transport calculations.

**Comment by arovinelli** [comment]
Some other insights

from the opt test with `--distributed-mesh` and `    `min_parallel = 3`  I receive the following

```
3d_auto_test_distributed: Writing Exodus 
3d_auto_test_distributed: Memory leak detected!
3d_auto_test_distributed: Compile in DEBUG mode with --enable-reference-counting
3d_auto_test_distributed: for more information
```

and the resulting exodus is actually missing nodes

```
3d_auto_test_distributed: ERROR: 
3d_auto_test_distributed:    *****************************************************************
3d_auto_test_distributed:               EXODIFF	(Version: 2.90) Modified: 2018-02-15
3d_auto_test_distributed:               Authors:  Richard Drake, rrdrake@sandia.gov           
3d_auto_test_distributed:                         Greg Sjaardema, gdsjaar@sandia.gov          
3d_auto_test_distributed:               Run on    2019/11/20   09:52:07 CST
3d_auto_test_distributed:    *****************************************************************
3d_auto_test_distributed: 
3d_auto_test_distributed: Reading first file ... 
3d_auto_test_distributed: Reading second file ... 
3d_auto_test_distributed:   FILE 1: sh_by_block_generator/gold/break_mesh_3D_auto_in.e
3d_auto_test_distributed:    Title: break_mesh_3D_auto_in.e
3d_auto_test_distributed:           Dim = 3, Blocks = 3, Nodes = 42, Elements = 8, Nodesets = 6, 
3d_auto_test_distributed:           Vars: Global = 0, Nodal = 0, Element = 0, Nodeset = 0, Sideset = 
3d_auto_test_distributed: 
3d_auto_test_distributed:   FILE 2: sh_by_block_generator/break_mesh_3D_auto_in.e
3d_auto_test_distributed:    Title: break_mesh_3D_auto_in.e
3d_auto_test_distributed:           Dim = 3, Blocks = 3, Nodes = 40, Elements = 8, Nodesets = 6, 
3d_auto_test_distributed:           Vars: Global = 0, Nodal = 0, Element = 0, Nodeset = 0, Sideset = 
3d_auto_test_distributed: 
3d_auto_test_distributed: exodiff: ERROR: .. Number of nodes doesn't agree.
3d_auto_test_distributed: 
3d_auto_test_distributed: exodiff: Files are different
3d_auto_test_distributed: 
3d_auto_test_distributed:  
3d_auto_test_distributed: ####
3d_auto_test_distributed: Tester failed, reason: EXODIFF
3d_auto_test_distributed: 
```
with


from the opt test with `--distributed-mesh` and `    `min_parallel = 2`  MOOSE crashes


> libMesh is still working on updating its interfaces to C++11 standards. Allocating a new node and adding it to the mesh is the way it's been done since libMesh was created. Now in this intermediate stage, it looks very weird since there's a mix of smart and non-smart pointer interfaces. The creation of a new Node through a smart pointer requires that we release the manager before adding it to the old (manually managed) interface. As far as I can tell, this is all correct. libMesh expects you to pass it a Node (or Element) through a pointer that it will then manage.

**Comment by permcody** [comment]
> I actually do not quite like the fact of libMesh having DistributedMesh and ReplicatedMesh.

It's true that you could use the algorithms in DistributedMesh in a replicated mode but they are heavier and slower, requires additional communication, etc. ReplicatedMesh serves most people very well and isn't likely to go anywhere anytime soon. Everything about the internals are simpler and more efficient for smaller problems. These two objects serve very different needs and simulation types.

**Comment by roystgnr** [comment]
> It's true that you could use the algorithms in DistributedMesh in a replicated mode but they are heavier and slower, requires additional communication, etc. ReplicatedMesh serves most people very well and isn't likely to go anywhere anytime soon. Everything about the internals are simpler and more efficient for smaller problems. These two objects serve very different needs and simulation types.

I was about to type this, except with more conflict-of-interest.  :-)  We could go through DistributedMesh bit by bit and make sure it always uses the replicated algorithms when it's in a replicated state, but that would still have more overhead from the underlying containers.  It would also make replicated meshes less useful as "training wheels", if changing a flag here or there could suddenly drop you into a full distributed state.

Case in point.  When you do ```new_node = Node::build(*current_node, mesh->n_nodes()).release()```, mesh->n_nodes() is very unlikely to be the id number you want.  To account for simultaneous asynchronous additions of both partitioned and unpartitioned nodes, DistributedMesh typically assigns ids in a sparse fashion (renumbering in prepare_for_use() to compress out the gaps), and if you're picking out your own ids then you need to either use that (e.g. leaving the id() unset and adding unpartitioned nodes in sync then partitioned nodes only on their processor) or at least avoid conflict with that (which can be as easy as storing an up-to-date max_node_id() before your modifications and using that as an offset, but the details depend on exactly what you're doing).

**Comment by lindsayad** [comment]
> Case in point. When you do `new_node = Node::build(*current_node, mesh->n_nodes()).release()`, mesh->n_nodes() is very unlikely to be the id number you want.

@roystgnr Are you suggesting that this could be the cause of the leak?

**Comment by roystgnr** [comment]
I'd guess so, but I still can't figure out how.  Most of the obvious ways that that could cause a leak are covered with libmesh_assert() - you can't just try to add_node on top of a different existing node, for example.  I assume someone's verified the leak in dbg or devel mode already?

**Comment by arovinelli** [comment]
All, I'm playing with what Roger suggested (at the moment I just added a shift of 50 and added a counter) and tests passe in both standard and debug mode with distributed options and different mpi ranks and no leak detected.
The question now is do I really need to have the same id for the new nodes on each cpu or not??

**Comment by lindsayad** [comment]
> Roger

You mean Roy?

> The question now is do I really need to have the same id for the new nodes on each cpu or not??

If I understand correctly, your code should either be:
```c++
if (current_node->processor_id() == mesh->processor_id())
{
  // This should use the Node copy constructor which will
  // assign the new node's processor ID to the current node's
  new_node = Node::build(*current_node).release(); 
  mesh->add_node(new_node);
}
```
or
```c++
  // Uses the Point and ID constructor. Let the DistributesMesh figure
  // out the ID and the partition for you
  new_node = Node::build(*current_node, DofObject::invalid_id).release(); 
  mesh->add_node(new_node);
```
@roystgnr can hopefully correct me if I'm wrong...

**Comment by roystgnr** [comment]
> do I really need to have the same id for the new nodes on each cpu or not?

The documentation says "Only [manually set id] in parallel if you are manually keeping ids consistent", and I'd love to stop at that, but of course it's not actually possible during distributed mesh refinement, so there are cheats possible.  What I do for AMR is apply temporary ids first and later use ```MeshCommunication::make_node_ids_parallel_consistent()```, which syncs the "canonical" ids on the owning processors to all the ghost nodes, identifying equivalent nodes by a same-processor-id element connected to them.  But that method relies on sync_node_data_by_element_id(), which is expensive enough that you'd want to avoid it if you can.

Sadly I don't think *either* of the methods @lindsayad pointed out will work in general.  In the first code, each processor will still be missing a ghosted copy of other processor's nodes.  In the second, you have to add unpartitioned nodes in sync (the same way you'd have to on a ReplicatedMesh) or their ids won't end up in sync either.

The ideal solution is to manually set nodes using some "canonical" ordering that doesn't require communication.  For instance, I recall writing mesh extrusion code to simply increment nodes' ids by max_node_id for each layer you extrude through, thereby making the new ids unique and independent of what processor sets them.  (although I can't recall, what app was that code in?)  In MOOSE the MeshCollectionGenerator works this way: by using max_node_id as an offset to copy_nodes_and_elements it guarantees canonical and unique ids there.  I'll see if I can find other examples.

**Comment by arovinelli** [comment]
@roystgnr and @lindsayad  thanks for the insight and all the reply. The only way I can see this working is to follow a synchronous process involving MPI communication which would read as follow

1. Each Rank should identify all the nodes that need to be replicated and constructing a map
2. One rank should grab all the information about the new nodes by all the other ranks, and create `canonical` ids for the new node
3. redistribute the entire information to all the ranks 
4. Each rank add nodes with conformal ID.

or 

1. use the same algorithm while keeping track of the added nodes and making sure they are unique 2. join the maps
3. renumber ids

**Comment by roystgnr** [comment]
Hmm... you know max_node_id.  And n_subdomains-1 should be the most copies you need of any node, so if max_node_id*n_subdomains fits in a dof_id_type then you shouldn't need any communication at this stage.  Can you assign
```
new_node_id = (connected_subdomain_id+1) * max_node_id + old_node_id
```
And then wait for the libMesh prepare_for_use to give you a dense numbering?

**Comment by arovinelli** [comment]
@roystgnr thanks for the idea
I can try and yes I can wait for prepare_for_use to do its job. A couple of questions:
1. where should I put prepare_for_use();
2. by max_node_id you the total number of node in the entire mesh i guess
3. what is the maximum for dof_id_type?

**Comment by roystgnr** [comment]
1. It's been too long since I edited any MOOSE mesh generators code, but IIRC there's already a prepare_for_use() down the line and you won't have to add it again.
2. Nope.  It's a separate MeshBase method.  If you have sparse numbering for any reason then it'll exceed the number of nodes in the mesh.
3. By default 4 bytes unsigned, so 4 billion ish.  MOOSE works with 8 byte dof_id_type too, but I'm not sure how many of your users are using 4 vs 8.

**Comment by arovinelli** [comment]
@roystgnr unfortunately I'm not able to find what is wrong with my changes
2 things happens:

1.  all tests in debug fails
2. when using the `--distributed-mesh` results depends on the number of CPU there is some random crash depending on the number of CPU used (test `break_mesh_by_blocks_generator_3d_auto_test_distributed` works with 3 CPU and crashes with two)

typical error messages are 
```
Assertion `!n->valid_id() || n->id() == _nodes.size()' failed.
2d_polycrystal_test: 
2d_polycrystal_test: [0] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/replicated_mesh.C, line 476, compiled Oct 22 2019 at 13:10:38
2d_polycrystal_test: Assertion `!n->valid_id() || n->id() == _nodes.size()' failed.
```

and
```
2d_splittrue_test: Assertion `this->comm().semiverify(obj ? &uniqueid : nullptr)' failed.
2d_splittrue_test: 
2d_splittrue_test: Assertion `this->comm().semiverify(obj ? &uniqueid : nullptr)' failed.
2d_splittrue_test: Assertion `this->comm().semiverify(obj ? &uniqueid : nullptr)' failed.
2d_splittrue_test: 
2d_splittrue_test: 
2d_splittrue_test: [2] buted_mesh.C, line 956, compiled Oct 22 2019 at 13:10:36
2d_splittrue_test: [0] buted_mesh.C, line 956, compiled Oct 22 2019 at 13:10:36
2d_splittrue_test: [1] buted_mesh.C, line 956, compiled Oct 22 2019 at 13:10:36
```

**Comment by roystgnr** [comment]
That 2d_polycrystal_test failure looks like you're trying to do this id cleverness on ReplicatedMesh as well, which unfortunately won't work - on ReplicatedMesh the id isn't just a mapping key, it's actually an index into a vector of pointers.  But there the fix is easy: just leave the id as invalid_id and as long as you're adding objects in sync you're fine.

The 2d_splittrue_test - I'm not sure without more context, but it looks like the new nodes' unique_id() values aren't in sync and our dbg test is complaining.  We might have to actually manually sync those before the prepare_for_use(), now that I think about it.  Let me see if I can replicate.

**Comment by roystgnr** [comment]
Yeah, I can replicate the problem, and I think it can be avoided.  Let me try a possible fix.

**Comment by roystgnr** [comment]
Well, I got a fix for that particular test at least.  Try hammering my fork of your branch and see if it works more generally?

**Comment by arovinelli** [comment]
@roystgnr Thanks for spending time on this.

unfortunately the solution is not general:

1. opt and replicated everything is ok
2. debug and replicated i have
```
2d_splittrue_test [min_cpus...] OK
3d_auto_test_distributed [skipped dependenc...] SKIP
3d_split_test .... FAILED (ERRMSG)
2d_polycrystal_test [min_cpus=3,OV...] FAILED (ERRMSG)
3d_polycrystal_test [min_cpus=3,OV...] FAILED (ERRMSG)
3d_auto_test ..... FAILED (ERRMSG)
2d_auto_test ..... FAILED (ERRMSG)
```

3. `distributed -- opt`the large tests fail (3d_polycrystal_test and 2d_polycrystal_test) hitting the same assertion
```
3d_polycrystal_test: [ 2] ***ASSERTION failed on line 243 of file /home/arovinelli/projects/petsc-3.10.5/arch-linux2-c-opt/externalpackages/git.parmetis/libparmetis/match.c:k >= firstvtx && k < lastvtx
3d_polycrystal_test: [ 2] 759 1138 -220624857 0 1
```

4. `distributed -- dbg` the large tests fail in the same way as opt distributed.

**Comment by roystgnr** [comment]
Ah, I see where I screwed up the replicated code path.  Easy fix, I just misunderstood one of our old APIs.

Not sure yet about the distributed code path, but I can at least reproduce the failures, let me look closer.

**Comment by roystgnr** [comment]
Okay, I at least see *what* the problem is; a side boundary id is getting generated differently on different processors.

**Comment by roystgnr** [comment]
Yup.  The invalid boundary_ids correspond to the same *name*, but they got assigned different boundary_id_type values.

**Comment by roystgnr** [comment]
Yeah, here we are:

```
BreakMeshByBlockGeneratorBase::findBoundaryNameAndInd
...
  // TODO need to be updated if distributed mesh is implemented
  // comments are left to ease implementation
```

Hope this gets you pointed in the right direction?  I'll push the correction to my branch that gets the id+unique_id problems solved, but I'm not going to have time to look at this boundary_id issue more today.

**Comment by arovinelli** [comment]
@roystgnr thank you so much for the help, I think I can handle it from here.
By the way I completely forgot about that comment that I wrote in the original mesh modifier.
I'll keep it from here
Thanks a lot again for the help and for time spent on this!!


> Yeah, here we are:
> 
> ```
> BreakMeshByBlockGeneratorBase::findBoundaryNameAndInd
> ...
>   // TODO need to be updated if distributed mesh is implemented
>   // comments are left to ease implementation
> ```
> 
> Hope this gets you pointed in the right direction? I'll push the correction to my branch that gets the id+unique_id problems solved, but I'm not going to have time to look at this boundary_id issue more today.

**Comment by roystgnr** [comment]
> I think I can handle it from here.

Yes, but the question is can I just drop the problem, unfixed, merely because I have more urgent work to do myself and don't want to be working before bed this weekend?  No.  The answer turns out to be no.

I've added another commit that I believe fixes the boundary id issue.

We then run into another failure, though; I'm looking into *that* now.

**Comment by roystgnr** [comment]
Okay, I think I see the remaining problem.  We're not always updating element node pointer links to duplicates of ghosted nodes.  If a ghosted node has to be replaced with a duplicate, but we don't actually see the elements on different subdomains that force its replacement, then we don't calculate a complete connected_elems set for that node, and if the incomplete set is of size 1 then we don't bother updating the node at all and now we're out of sync.

Nasty.  What's especially scary is that it's possible for even the *owner* of a node not to know it needs duplication, if the owner sees it only attached to local and ghost elements from subdomain A but doesn't see an attached remote element of subdomain B.  So I think on each processor we'll need to first build connected_elems sets for ghost nodes, then push_parallel_vector_data to their owners, use that push to generate and update connected_elems sets for nodes we own, then pull_parallel_vector_data to get those updated sets for our ghost nodes, then *finally* we can use the updated sets (alongside sets built on the fly for nodes we own but haven't been queried about) to build duplicate nodes.

Needless to say this should only be done ```if(!is_replicated)```.

That's going to be quite difficult, but hopefully you want to take first crack at it anyway?  In theory I'm on vacation starting in 25 minutes, and in practice even when I'm getting work done it'll probably be the work I just procrastinated because your problem was more interesting.  :-D

**Comment by arovinelli** [comment]
@roystgnr Dear Roy, thanks a lot for spending so much time on this.
I'll work on creating the required list as soon as I can. It might time some time as I'm a bit rusty with MPI, but is shouldn't be impossible. just bit of pain organizing the data structure and making sure we get the complete list (as you mentioned we need to start from elements)
In the meanwhile I reiterate my deep appreciation for your help. and I wish you wonderful vacations.

Andrea

**Comment by roystgnr** [comment]
Take a look at the parallel_sync.h usage in libMesh?  You should be able to do this, using the super-efficient algorithms @friedmud now has backing my API there, without a single raw MPI call.  Now if you're rusty with C++11 lambda expressions, on the other hand, *that* might take some time to get up to speed.  ;-)

**Comment by arovinelli** [comment]
@roystgnr thanks for the suggestion I'll take a look at that.
In the meanwhile I think I solved the opposite problem (joining faces), which should be much easier in a distributed setting (see this branch [JoinByFaces](https://github.com/idaholab/moose/compare/next...arovinelli:JoinFaces)

**Comment by arovinelli** [comment]
@permcody I'm starting to be confused here. According to the documentaion `--split`-mesh` will make the split acting on a replicate mesh. However, you are telling me that the modiifer must be able to run in `Distributed` mode.
Does this mean that I should first `--split-mesh` and when loading it apply modifers?
or should I split an already modifed mesh?

**Comment by permcody** [comment]
Yeah - there are lots of confusing pieces here. ReplicatedMesh and
DistributedMesh are completely different objects so that's what I mean when
I say that your generator should ideally be able to work with each object.
Note that many of the existing generators definitely do NOT work with
DistributedMesh. That being said, it's possible and normal for
DistributedMesh objects to not be distributed during the generation phase.
It's fairly normal to build up a mesh simultaneously on all processors and
then partition it up and delete the remote pieces later. The confusion is
that we have a nasty mix of cases in the framework and we don't have clear
direction on when you can expect your mesh to be fully available or not
(regardless of the type). For instance, one of the most commonly used
generators GeneratedMeshGenerator (terrible name) calls libMesh to generate
the mesh and libMesh automatically calls "prepare_for_use()", which removes
the remote elements before handing the mesh back to MOOSE for use. If you
have another generator that runs after that step, you won't have access to
the remote pieces... This is why you are struggling to make your piece
work. In other custom generators that build up the mesh using MOOSE logic,
we may or may not call prepare_for_use() meaning that you will have access
to the remote pieces. This makes the whole state of the generator system
difficult to navigate. I can say that we are aware of these issues and it's
on our list of things to cleanup and fix so that it's much easier for
developers to work with.

Split mesh - yes, split mesh is designed to work with ReplicatedMesh only.
Note that this requirement may be dropped in the future if we can guarantee
that the full mesh is available on all processors at the time of the split
but based on my paragraph above - we aren't there yet. What happens in
Split Mesh is you build up a mesh and run it through one or more
generators, then when we are all done, we run a partitioner (and all of the
active relationship managers on the mesh) and write out the pieces. The
type doesn't really matter. The problem is that it's important that all
processors have access to the full mesh since we can run "serial"
partitioners on it to perform the split.

I realize that this is all very complicated so my current advice should be
really don't worry about all of these complications just yet because the
framework could use some serious improvements here. Perhaps you really
should only worry about the ReplicatedMesh part since if you have that you
can run your problem with ReplicatedMesh and DistributredMesh if you use
the split-mesh capability. Note that you won't be able to run your problem
with DistributedMesh mode directly (which is admittedly odd). Not sure if
that clears anything up or just adds to the confusion.

**Comment by arovinelli** [comment]
@permcody thanks for the exhaustive reply.

If I understand what you are saying correctly, running the split in a serial fashion, (`mpiexec -n 1 input.i --split-mesh 50 -split-file foo`) should solve all my problems beacuase the whole mesh should always be available (or maybe not)?
Is this correct?

**Comment by permcody** [comment]
@arovinelli - Yes

**Comment by arovinelli** [comment]
@permcody I did try the above for a simple case and it partially works.

However it misses the most important feature of the `breakmeshbyblock` because the connectivity at the interface is lost when performing `--split-mesh`.
(the all idea of this mesh generator is to split the mesh while retaining the connectivity between the splitted blocks)
So now I'm asking for ideas.

> @arovinelli - Yes

**Comment by permcody** [comment]
This is a solved problem. Every object that needs nonlocal information
should be using the RelationshipManager system to advertise what it needs.
Then you'll have the additional elements preserved.

**Comment by arovinelli** [comment]
@permcody ok, seems promising, how do I do that (examples?)?

**Comment by permcody** [comment]
Hopefully there is enough here. This is of course another very complex
system:
https://www.mooseframework.org/newsletter/2018_07.html

**Comment by arovinelli** [comment]
Either I'm doing something wrong with the RM (my setup should be exactly the same of DG kernels and I put it it breakmesh by block) or the neighbor map (`elme.set_elem_neighbor(face, neighbor)`) is reconstructed on the fly when using the flag `--use-split` and not saved in cpr file, becasue it doesn't work. If the neighbor map is reconstructed on the fly then I need another strategy or break the mes after the split.

**Comment by permcody** [comment]
Are you seeing the printout from the RM system on your console when splitting and reading back in? 

Important note if this wasn't clear before. The RM system is a consumer based advertising system. That is, the supplier of an object doesn't normally know what may or may not be needed downstream, but as an object creator you generally know when you'll need extra information. (e.g. you are a DG kernel so you need to be able to access neighbor information everywhere. The RM should be "published" on the consumer side. I'm not sure where you put the RM in your system but it should be in your DG kernel or whatever other object you are using and that should be in your input file when you split (it might seem insignificant, but it's vitally important). The system will only preserve information if some object is saying it needs it later.

**Comment by arovinelli** [comment]
@permcody I set up a illustrating this problem, so we can be on the same page to discuss this.

[branch](https://github.com/arovinelli/moose/commit/9536a6084c39fa72f4c1152d1d9d06700cdcf01e)

This is branch is just using a previous example with the `penaltyinterfacekernel` and `breakmeshbyblock`.
I also add the RM stuff to the `penaltyinterfacekernel`.
If you run it serially everything works fine.
if you presplit, and load the split mesh (see test naemd `use_split_mesh`) then it doens't produce the same result as in the serial case. Can you take a look if I added the RM properly?

**Comment by permcody** [comment]
Excellent - one question, why two different input files for splitting and reloading? That shouldn't be the cause of your issue since the Penalty object exists in both. Just curious.

**Comment by arovinelli** [comment]
becuase of #14440 ,and apparently i can't to
```
[Mesh]
   type = generatedmesh
   ....
[]
[MeshModifiers]
whatever
[]
```

> Excellent - one question, why two different input files for splitting and reloading? That shouldn't be the cause of your issue since the Penalty object exists in both. Just curious.

**Comment by permcody** [comment]
Fair enough - it seems you have always been on the bleeding edge. Have you though about moving to Idaho? 😄

**Comment by arovinelli** [comment]
@permcody  well, you'll have to convince more than person :smile:
In the meanwhile what if I intercept the mesh here https://github.com/idaholab/moose/blob/f869dae4807133c6023b2290612cd181f8358f75/framework/src/actions/SetupMeshAction.C#L270-L274

and add something like

```
 if (_app.isUseSplit() && someflag);
    _mesh= customGeneratorToJoinMeshByFaces(_mesh)
```

is this going to work?

**Comment by permcody** [comment]
Well that sounds like a horrible hack to me 😄.

Sorry, I'm context switching a lot today (like everyday) so I'm not sure if I understand which problem you think this will fix. I gather you are trying to fix the "reload" problem where you aren't seeing the necessary neighbor information reflected from the pre-split mesh. If so, that might do it, but I really don't see a good way of checking that in. Perhaps you were just thinking about a good local hack?

BTW - That same chunk of code is probably the reason for #14440, but I'd have to dig into it more.

**Comment by arovinelli** [comment]
@permcody, I know this is a horrible hack, the problem that I believe is going to fix is the fact that we loose the connectivity because such information is not saved when performing the split and connectivity is reconstructed when reloading the mesh. Now, because we broke the mesh the interface node will not share the same nodes anymore and therefore will not be neighbor, doesn't matter how many ghost layers you use. 
So i can see two ways to fix this issue:
1. save extra information in the cpr file and use them when reloading the `broken` and `split` mesh, or
2. add some kind of final preprocessor after the split mesh has been loaded (as far as I understand it is not possible now to  use `MeshGenerators` after reload of a split mesh)

The latter option seems the closer to the MOOSE spirit, however I'm not sure how to make it happen. Maybe is as easy as add a MeshGenrator with some special `input` name that will capture the mesh, reset the neighbors of such faces (I already have some code for this ) and then pass it to the solver.

**Comment by permcody** [comment]
> 1. save extra information in the cpr file and use them when reloading the broken and split mesh, or

What is this extra information? I'm always open to saving more information but it seems like a mesh is a mesh, right? What's important in terms of splitting is to make sure you have enough "remote" elements saved off in the pieces.

> add some kind of final preprocessor after the split mesh has been loaded (as far as I understand it is not possible now to use MeshGenerators after reload of a split mesh)

The design of split mesh was that you'd rerun the same input file for splitting and running. If that's your starting point, we didn't intend for the addition of new MeshGenerators so yes, it's not currently possible to add more. However, this seems like a workaround. I'd like to understand more about why information is being lost.

> The latter option seems the closer to the MOOSE spirit

I don't think so (see my discussion above). Splitting is just for splitting, not for restart or recover or any of those other things that are more in-between where we do support adding additional objects.

**Comment by arovinelli** [comment]
@permcody
Ok let's start from what is happening in this generator
let's say you start with a mesh like below (2 Quad4 elments ) 
1   2    3
o---o---o
|     |     |
o---o---o
3    4    6 
this will give and element to node map like
e1 = {1,2,3,4}
e2 =  {2,3,4,6}
whit this information if you have enough ghosting layers (1 layer in this case) you can reconstruct the neighbor (wahtever is set from `elem.set_neighbor()`) connectivity because e1 and e2 shares nodes 2 and 4.

When you apply `BreakMeshByBlockGenerator` you end up with
1   2    7    3
o---o    o---o
|     |     |     |
o---o    o---o
3    4    8    6
so now your element to node map is 
e1 = {1,2,3,4}
e2 =  {7,3,8,6}
if I want to find neighbors now I can't because e1 and e2 do not share nodes anymore
Now if you have followed me until here everything the following is trivial:
when you reopen the mesh and try to setup the neighbors from a broken mesh, you we all have all the required element layers because of ghosting, however you will not be able to find that e1 is neighbor of e2.

The missing information is actually the element,<face, neighbor> map.
You can either save it in the cpr file and reload it (preferred option for my, will also solve restart with broken meshes and large dispalcements) or you can reconstruct on the fly (which is what is done now and is missing the neighbors).

I hope this clarify the issue

One way to solve this might be save and reloading all the neighbor information on boundaries (if present)

**Comment by permcody** [comment]
Well that makes perfect sense and and explains why the built-in relationship manager that you choose fails to work here (e.g. no shared nodes).

Possible solutions:
1. Build a custom relationship manager that can recover the "disconnected" element information from the mesh. Sorry I haven't read through all the code, but I assume that the nodes are co-located? If so, it should be possible to recover this information using a RelationshipManager. 

2. Return back to trying to make your MeshGenerator work with DistributedMesh

3. Wait for us to fix all this by coming up with a sane workflow where you don't have to worry about parts of your mesh getting deleted before you are ready for it to be deleted.

4. Run with ReplicatedMesh, but I assume you wouldn't be doing any of this if you were satisfied with that approach.

5. Write out more information to a file somewhere and read it back in, not too bad, but essentially building your own "RelationshipManger", right?

Option 3 might be the best, but Option 1 isn't bad either. Creating your own Relationship Manager isn't trivial. It basically involves returning the elements "connected" to an element (for every element), where connected is your custom logic. In the future, we plan to use RelationshipManagers to improve how we implement ghosting for things like contact or mortar interfaces (i.e. relationships that aren't connected at all, similar to your case here).

OK, here's a bigger question: What projects are you working on and who is funding you? Would it benefit you to visit the lab and work with our team more directly at some point? Maybe we could talk through a lot of your issues, give you our vision for our current design, or future directions, etc.?

**Comment by lindsayad** [comment]
> mortar interfaces

We use a relationship manager right now for mortar; see the [`AugmentSparsityOnInterface`](https://github.com/idaholab/moose/tree/next/framework/src/relationshipmanagers/AugmentSparsityOnInterface.C) object. As of this very moment it only works for replicated mesh, but I'm going to go fix that right now!

**Comment by lindsayad** [comment]
While trying to get mortar set-up with `DistributedMesh` I realized that interaction of `MeshGenerators` with geometric `RelationshipManagers` is totally broken. See #14451

**Comment by arovinelli** [comment]
> Well that makes perfect sense and and explains why the built-in relationship manager that you choose fails to work here (e.g. no shared nodes).
> 
> Possible solutions:
> 
>     1. Build a custom relationship manager that can recover the "disconnected" element information from the mesh. Sorry I haven't read through all the code, but I assume that the nodes are co-located? If so, it should be possible to recover this information using a RelationshipManager.

Yes node are colacated at the beginning, however this option will prevent restart if a mechanical simulataion with large strain is performed. Restart is one of my goal.
 
>     2. Return back to trying to make your MeshGenerator work with DistributedMesh
I thought about this option too, but is not viable as I realized I utilized a lot of meshgenerators that are not compatible with distributed mesh. While this is something to do in the long term it's not my number one priority at this point.

>     3. Wait for us to fix all this by coming up with a sane workflow where you don't have to worry about parts of your mesh getting deleted before you are ready for it to be deleted.
It would be nice but I need to fix this quickly as I have a huge simulation to run (see next answer)
> 
>     4. Run with ReplicatedMesh, but I assume you wouldn't be doing any of this if you were satisfied with that approach.
This is not viable as my plan is to run a 40Milion Dof simulation on our cluster (the mesh it'self is 40GB). The only viable option at the moment seems to be `--split-mesh`
> 
>     5. Write out more information to a file somewhere and read it back in, not too bad, but essentially building your own "RelationshipManger", right?
This is probably the best option, and the one I'm going to pursue. This will not only allow to have the relationship fixed, but it would also allow restart. This would imply having a persistent map of `initially coincident nodes` and of `face-face neighbors`
> 
> 
> Option 3 might be the best, but Option 1 isn't bad either. Creating your own Relationship Manager isn't trivial. It basically involves returning the elements "connected" to an element (for every element), where connected is your custom logic. In the future, we plan to use RelationshipManagers to improve how we implement ghosting for things like contact or mortar interfaces (i.e. relationships that aren't connected at all, similar to your case here).
> 
@lindsayad said there something similar to what I need for option 5 in mortar some I'm going to take a look at this.


> OK, here's a bigger question: What projects are you working on and who is funding you? Would it benefit you to visit the lab and work with our team more directly at some point? Maybe we could talk through a lot of your issues, give you our vision for our current design, or future directions, etc.?

Cody, thanks for the invitation, I really appreciate it. We can talk about this offline.

**Comment by permcody** [comment]
Any time invested in a "write out extra information" yourself solution can hopefully be transferred to building your own RelationshipManager. Before you go too far down this path, take a look at the existing relationship managers and see if you can  come up with something. If you go that direction you'll avoid all of the pain of File I/O at scale. 

To repeat, the extra information you need to write is precisely the logic you need to encode in a RelationshipManager.

**Comment by arovinelli** [comment]
> Any time invested in a "write out extra information" yourself solution can hopefully be transferred to building your own RelationshipManager. Before you go too far down this path, take a look at the existing relationship managers and see if you can come up with something. If you go that direction you'll avoid all of the pain of File I/O at scale.
Of course the idea is to integrate it, let's say option 5 but using the relationship manger and all the required tools. I'm looking at  [AugmentSparsityOnInterface](https://github.com/idaholab/moose/blob/next/framework/src/relationshipmanagers/AugmentSparsityOnInterface.C) and [MortarConstraint](https://github.com/idaholab/moose/blob/next/framework/src/constraints/MortarConstraint.C) and trying to figure out what I can use reuse
> 
> To repeat, the extra information you need to write is precisely the logic you need to encode in a RelationshipManager.

**Comment by permcody** [comment]
Another good reference would be to look in libMesh and see the documentation on "ghosting functors". I'm too lazy to find the exact spot but you can find it if you look for those keywords.

**Comment by arovinelli** [comment]
@lindsayad and @permcody Ok I think I got the idea. As far as I understand the RM will always be executed after the generation phase is terminated, which means that in my case I can save information about coupled node/element/elementfaces (this is still not all clear to me do I need nodes?) inside the `BreakMeshByBlockGenerator` retrieve information in the RM and incorporate the RM inside my `InterfaceKernel`. I'm still confused about how this is going to help to reconstruct the which side of which element is neighbor to whom for the interface kernel (maybe MOOSE takes care of it automatically?)

**Comment by lindsayad** [comment]
> Another good reference would be to look in libMesh and see the documentation on "ghosting functors". I'm too lazy to find the exact spot but you can find it if you look for those keywords.

See the Detailed Description [here](https://libmesh.github.io/doxygen/classlibMesh_1_1GhostingFunctor.html)

**Comment by permcody** [comment]
That's for quoting me being lazy and then proving that you're not :)

**Comment by arovinelli** [comment]
@roystgnr
following your suggestion I constructed the new nodelist and everything works fine for a replicated mesh. Now I'm trying to move to replicated mesh and I want gather all the nodelist cosntructed by each process assemble them and broadcast the complete list to all process. I'm trying to use ` Parallel::pull_parallel_vector_data` that, if I understand correctly does a gather + broadcast of a vector data structure creating a map <pid, data_of_pid> but something is not working (only rank 0 does stuff)

here is what I'm trying. Am i doing something wrong?

```
  if (!mesh->is_replicated())
  {
    /* need to do communication
     we can only use vectors and vectors of vectors for communication so we start unrolling the
     node_to_domains_map */
    std::map<processor_id_type, std::vector<dof_id_type>> data_to_send, received_data;
    data_to_send.clear();
    received_data.clear();

    // fill the vector fo data to send
    for (auto it_node : node_to_domains_map)
      for (auto it_sub : it_node.second)
      {
        data_to_send[mesh->processor_id()].push_back(it_node.first);
        data_to_send[mesh->processor_id()].push_back(it_sub);
      }

    std::cout << " Rank " << mesh->processor_id() << " DataToSendSize "
              << data_to_send[mesh->processor_id()].size() << "!!!!!!!!!!!!!!!!!!! " << std::endl;

    // compose replies from each cpu
    auto gather_data = [](processor_id_type pid,
                          const std::vector<dof_id_type> & query,
                          std::vector<dof_id_type> & response) {
      std::cout << "gather_data pid " << pid << std::endl;
      response = query;
    };

    // put together all the responses in a map
    auto act_on_data = [&received_data](processor_id_type pid,
                                        const std::vector<dof_id_type> & query,
                                        const std::vector<dof_id_type> & response) {
      std::cout << "act_on_data pid " << pid << std ::endl;
      received_data[pid] = response;
    };

    dof_id_type * ex = nullptr;
    Parallel::pull_parallel_vector_data(mesh->comm(), data_to_send, gather_data, act_on_data, ex);

    
      // print some output
      for (auto const & x : received_data)
        std::cout << " Rank " << mesh->processor_id() << "  received data from Rank " << x.first
                  << " DataSize " << x.second.size() << "!!!!!!!!!!!!!!!!!!! " << std::endl;
    mesh->comm().barrier();
    }
```

**Comment by lindsayad** [comment]
I just don't think this is going to work at all with `DistributedMesh` until #14451 is resolved. There's no guarantee that the elements you want to work with in your generator will still be around after the initial mesh generator is run and executes `prepare_for_use`

**Comment by arovinelli** [comment]
@lindsayad That's the all point of adding the communication part. Even if element have already been deleted, each process can still construct it's own map. Then I will gather all the maps, assemble them and construct a new unique node_ids subdomain map, broadcast the entire map to all the cpus,  and add node as needed on each process, Why should this not work?

**Comment by arovinelli** [comment]
@lindsayad actually,  gather + assemble + brodcast is enough,  as each process can construct the entire map independently

**Comment by roystgnr** [comment]
Those clears are unnecessary on just-constructed objects.

But more importantly:  The index you want for data_to_send isn't mesh->processor_id(); that'd only ever send to yourself!  You index the queries by which remote processor you need to ask for data from.

**Comment by lindsayad** [comment]
> That's the all point of adding the communication part. Even if element have already been deleted, each process can still construct it's own map. Then I will gather all the maps, assemble them and construct a new unique node_ids subdomain map, broadcast the entire map to all the cpus, and add node as needed on each process, Why should this not work?

No, that sounds good!

**Comment by arovinelli** [comment]
@roystgnr 
I'm doing some step forward but I believe I'm still missing something.
Just to recap.
first i flatten the data that i want to send
```
    std::vector<dof_id_type> flattened_data;
    // faltten data to send
    for (auto it_node : node_to_domains_map)
      for (auto it_sub : it_node.second)
      {
        flattened_data.push_back(it_node.first);
        flattened_data.push_back(it_sub);
      }
```
Then I populate the data to send map. because I want all cpus to send their portion to all other cpus I copy the same data on all map values:
```
    std::map<processor_id_type, std::vector<dof_id_type>> data_to_send, received_data;
    for (processor_id_type i = 0; i < mesh->n_processors(); i++)
      data_to_send[i] = flattened_data;
```

My gather and action functor are 

```
    // compose replies from each cpu
    auto gather_data = [](processor_id_type pid,
                          const std::vector<dof_id_type> & query,
                          std::vector<dof_id_type> & response) {
      response = query;
      std::cout << "gather_data pid " << pid << " response_size " << response.size()
                << " query_size " << query.size() << std::endl;
    };

    // put together all the responses in a map
    auto act_on_data = [&received_data](processor_id_type pid,
                                        const std::vector<dof_id_type> & query,
                                        const std::vector<dof_id_type> & response) {
      received_data[pid] = response;
      std::cout << "act_on_data pid " << pid << " response_size " << response.size()
                << " query_size " << query.size() << " received_data size " << received_data.size()
                << std::endl;
    };

and my pull is 
```
    dof_id_type * ex = nullptr;
    Parallel::pull_parallel_vector_data(mesh->comm(), data_to_send, gather_data, act_on_data, ex);
```

To check that results are what I want (my received_data[pid] should a have a size equal to the data sent from the query generated by procesdor <pid> )
I print 
```
    // print some output
    for (auto const & x : received_data)
      std::cout << "Rank " << mesh->processor_id() << "  received data from Rank " << x.first
                << " DataSize " << x.second.size() << "!!!!!!!!!!!!!!!!!!! " << std::endl;
```

however the outcome i not as expected because processor 0 receive always data of the same size from all processes (see below). i suspect something is wrong either on how I gether the data or on how I assemble the data (I'm never using a pid in the gather which is suspcicious)

```
gather_data pid 0 response_size 26 query_size 26
gather_data pid 8 response_size 0 query_size 0
gather_data pid 1 response_size 26 query_size 26
gather_data pid 2 response_size 16 query_size 16
gather_data pid 3 response_size 0 query_size 0
gather_data pid 4 response_size 16 query_size 16
gather_data pid 5 response_size 0 query_size 0
gather_data pid 6 response_size 26 query_size 26
gather_data pid 7 response_size 0 query_size 0
act_on_data pid 0 response_size 26 query_size 26 received_data size 1
act_on_data pid 7 response_size 26 query_size 26 received_data size 2
act_on_data pid 8 response_size 26 query_size 26 received_data size 3
act_on_data pid 1 response_size 26 query_size 26 received_data size 4
act_on_data pid 2 response_size 26 query_size 26 received_data size 5
act_on_data pid 3 response_size 26 query_size 26 received_data size 6
act_on_data pid 4 response_size 26 query_size 26 received_data size 7
act_on_data pid 5 response_size 26 query_size 26 received_data size 8
act_on_data pid 6 response_size 26 query_size 26 received_data size 9
 Rank 0  received data from Rank 0 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 1 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 2 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 3 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 4 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 5 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 6 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 7 DataSize 26!!!!!!!!!!!!!!!!!!! 
 Rank 0  received data from Rank 8 DataSize 26!!!!!!!!!!!!!!!!!!! 
```

Thanks in advance

**Comment by arovinelli** [comment]
@roystgnr I think i figure it out but I'm not sure this is the supposed way to go

I simply fill my received data in the gather functor and do nothing on the acting functor as:
this should be full gather from and to all

```
    // compose replies from each cpu
    auto gather_data = [&received_data](processor_id_type pid,
                                        const std::vector<dof_id_type> & query,
                                        std::vector<dof_id_type> & response) {
      received_data[pid] = query;
    };

    // put together all the responses in a map
    auto act_on_data = [&received_data](processor_id_type pid,
                                        const std::vector<dof_id_type> & query,
                                        const std::vector<dof_id_type> & response) {};
```

**Comment by roystgnr** [comment]
Are you sure every single processor needs every bit of data from every single other processor?  That's not going to scale well.  I guess correctness is the more urgent concern though.

If you don't actually have a query-and-reply situation going on here, then you want to use the push_parallel_vector_data API rather than the pull; no point in doing a two-step communication when one of those steps is a no-op.

**Comment by arovinelli** [comment]
@roystgnr  
> Are you sure every single processor needs every bit of data from every single other processor? That's not going to scale well. I guess correctness is the more urgent concern though.
> 

At least one processor need to know everything about the boundary at some point in time to generate unique ids for new nodes on all the mesh.

So ideally what I want to do is
- let each processor figure out it's own data (which is already done)
- give all the data to one processor 0 (this should be a push)
- now processor 0 should do some manipulation
- once processor 0 is done with the manipulation should send back the assembled data to all the other processor (is this another push??)

But I'm starting to get confused on how the `push_parallel_vector_data` and the `pull_parallel_vector_data` work. My understanding is that I should be able to do all of the above with a single `pull_parallel_vector_data` in which the `gather_data` as below


```
    auto gather_data = [&received_data](processor_id_type pid,
                                        const std::vector<dof_id_type> & query,
                                        std::vector<dof_id_type> & response) {
      received_data[pid] = query;
    };
```

and the `act_on_data` some assembling procedure saving stuff somewhere but I'm not sure how to do this.
Is this correct or there is some design failure here (beside optimization which will be a later problem)

**Comment by permcody** [comment]
@arovinelli - If the *only* reason that you need communication is due to unique_id generation, then you don't need communication. You can always come up with a scheme where each processor can independently number things in such a way as not to conflict with other processors. A simple scheme would be to number things using an "n_processors" stride (e.g. if you are running on 10 processors, then the first processor's IDs would all be divisible by 10 and so on). You can offset that based on the existing numbering or whatever, but such a scheme could avoid parallel communication.

**Comment by arovinelli** [comment]
> @arovinelli - If the _only_ reason that you need communication is due to unique_id generation, then you don't need communication. You can always come up with a scheme where each processor can independently number things in such a way as not to conflict with other processors. A simple scheme would be to number things using an "n_processors" stride (e.g. if you are running on 10 processors, then the first processor's IDs would all be divisible by 10 and so on). You can offset that based on the existing numbering or whatever, but such a scheme could avoid parallel communication.

Unfortunately is not the only reason, when we are adding node we need to add them with a consistent id between cpus (so someone will need to make a map), and if you want this to work with a big mesh (millions of elements and hundreds of domains) you can't simply use a scheme as you will reach out of maximum numbering pretty easy. Also even if with quad elements it might work, the situation is much worse when using tets, when you have an unknown number of domains converging at one corner. Anyway at this point this point this becoming a metter of understanding how to properly use the parallel_sync, for which there is not a good documentation beside few lines in the doxygen

**Comment by arovinelli** [comment]
@permcody BTW how do I enable std::cout to screen for non rank0 processes??

**Comment by permcody** [comment]
> how do I enable std::cout to screen for non rank0 processes??

`--keep-cout`, however you will almost certainly also want `--redirect-stdout` as well. Note that when you use the latter option, you'll lose all output to your screen but you'll have sane output in several separate files.

**Comment by roystgnr** [comment]
> how do I enable std::cout to screen for non rank0 processes??

By default std::cerr still goes to the screen, however on large processor counts this is still pretty useless due to interleaving and redirecting to files is tempting.

**Comment by roystgnr** [comment]
> when we are adding node we need to add them with a consistent id between cpus (so someone will need to make a map), and if you want this to work with a big mesh (millions of elements and hundreds of domains) you can't simply use a scheme as you will reach out of maximum numbering pretty easy

All true, but there's still got to be a better solution than "do everything in serial".

Typically in these sorts of situations what I do is come up with a rule that enforces a unique owner processor on each new datum, then a rule for that owner to choose the id.  For duplicate nodes it seems easy enough to pick the owner: whoever owns the node you're duplicating picks it.  And the rule is easy too: just add_node with that ownership so that DistributedMesh will choose a unique id for you.  So I still do think my original algorithm idea would mostly work:

...first build connected_elems sets for ghost nodes, then push_parallel_vector_data to their owners, use that push to generate and update connected_elems sets for nodes we own, then pull_parallel_vector_data to get those updated sets for our ghost nodes, then finally we can use the updated sets (alongside sets built on the fly for nodes we own but haven't been queried about) to build duplicate nodes.

Except that:
It'd probably be better to keep track of connected_subdomains, to have less data to push.
Instead of just doing a pull_parallel_vector_data for updated connected_subdomains, we'd want to simply pull for newly-created node ids, because after the connected_subdomains have been pushed, each node's owner should have enough data to create the duplicate nodes.

> Anyway at this point this point this becoming a metter of understanding how to properly use the parallel_sync, for which there is not a good documentation beside few lines in the doxygen

Yeah, this is my fault, I apologize.  It doesn't help that we're using templated container types for flexibility either, so it isn't even as self-documenting as it could be there.

For push_parallel_vector_data, you create a ```map<vector<stuff>> data``` (using in most cases unordered_map) on each processor p, then the push sends each ```data[p2]``` to the corresponding processor p2, which then calls your ```act_on_data(p, the_data_from_p)``` to decide what to do with it.

For pull_parallel_vector_data, you fill each ```queries[p2]``` (where queries is again a ```map<vector<stuff>>```, but now what p2 does with that data is calls ```gather_data(p, the_query_from_p, the_answer_to_send_back)``` to let you come up with a response, and then when those responses are sent back to p it calls ```act_on_data(p2, the_query_to_p2, the_answer_sent_back)``` for you to act on those queries.

So in the case of the push, "stuff" might be something like a pair<

**Comment by roystgnr** [comment]
So for the first push, your ```stuff``` might be ```vector<dof_id_type>```, and each ```data[p2][i][0]``` would be the id of a ghost node owned by p2, and each ```data[p2][i][j]``` for positive j would be a subdomain id of an element owned by p which is connected to that ghost node.

For the pull ... well, this gets harder, because the queries don't know how *much* data they're querying, since each ghost node might have any number of duplicates.

The ```stuff``` might be ```vector<dof_id_type>```, and each ```queries[p2][i]``` would be the id of a ghost node owned by p2.  Then ```the_answer_to_send_back[i]``` could be a ```std::vector<std::tuple<dof_id_type, unique_id_type, subdomain_id_type>>``` where (when non-empty!) each entry would be the id+unique_id to use for a new node creation, plus the subdomain_id to which that node would be attached.  You'd look at your own connected_elems for that ghost node, and for each you'd check the subdomain to see if it needed to attach to a newly created node, then if so do so.

**Comment by roystgnr** [comment]
Or, if you really wanted to be communications-efficient, you could save the data from the first push (because it's got all the data you'd be querying in the second push anyway), and then do a second push based on that.  That way you'd have 2 communication steps instead of 3 (a pull is basically 2 pushes internally), plus you would be able to send only the data for the subdomains each processor has connected to a ghost node, rather than for every subdomain *any* processor sees connected to that node.

**Comment by permcody** [comment]
@roystgnr just put us over 100 comments on this PR. How much longer until [Godwin's Law](https://en.wikipedia.org/wiki/Godwin%27s_law) kicks in? 😄

**Comment by arovinelli** [comment]
@roystgnr thanks a lot for all the time spent here and for the detailed explanation about `parallel_sync`. It really helped. It also possible that there might be more efficient ways for getting out with new unique node ids, however at this point i want to focus in makeing this generator work rather than performance. Then, once everything is working properly we could come out with more efficient communication/node numbering strategies.

**Comment by arovinelli** [comment]
> @roystgnr just put us over 100 comments on this PR. How much longer until [Godwin's Law](https://en.wikipedia.org/wiki/Godwin%27s_law) kicks in? smile

Hopefully I'll be able to fix this thing with only other few comments :smile:

**Comment by lindsayad** [comment]
> The documentation says "Only [manually set id] in parallel if you are manually keeping ids consistent",

@roystgnr where is this documentation?

**Comment by lindsayad** [comment]
> > The documentation says "Only [manually set id] in parallel if you are manually keeping ids consistent",
> 
> @roystgnr where is this documentation?

@roystgnr where is the documentation guiding when to manually set ids?

**Comment by roystgnr** [comment]
It's in the doxygen comments for MeshBase::add_point and MeshBase::add_elem

**Comment by roystgnr** [comment]
But that's *when* - I'm not sure where we've got proper documentation as to *how*.  I remember writing it out at least once but that might have been on libmesh-users or libmesh-devel.

**Comment by lindsayad** [comment]
Ah ok, I was looking at add_node and didn't think to look at add_point

**Comment by lindsayad** [comment]
@jiangwen84 can you comment or put an 'x' on whether the first bullet in your original post has been fixed? I want to close this issue in favor of a couple more specific, newer issues, but I want to know what we've addressed already first

**Comment by jiangwen84** [comment]
Yes,  I fixed that.

**Comment by lindsayad** [comment]
Great, thanks!

**Comment by lindsayad** [comment]
Closing in favor of more specific issue #21154. I think issues have gotten too long when github forces us to expand comments
